java/core/src/main/java/co/worklytics/psoxy/storage/impl/CSVFileHandler.java

```diff
@@ -47,29 +49,45 @@ public byte[] handle(@NonNull InputStreamReader reader, @NonNull Sanitizer sanit
                 .stream()
                 .map(Rules.Rule::getCsvColumns)
                 .flatMap(Collection::stream)
-                .collect(Collectors.toSet());
+                .map(String::trim)
+                .collect(Collectors.toCollection(() -> new TreeSet<>(String.CASE_INSENSITIVE_ORDER)));
 
         Set<String> columnsToPseudonymize = options.getRules()
                 .getPseudonymizations()
                 .stream()
                 .map(Rules.Rule::getCsvColumns)
                 .flatMap(Collection::stream)
-                .collect(Collectors.toSet());
-
-        String[] header = records.getHeaderMap().keySet()
+                .map(String::trim)
+                .collect(Collectors.toCollection(() -> new TreeSet<>(String.CASE_INSENSITIVE_ORDER)));
+
+        // headers respecting insertion order
+        // when constructing the parser with ignore header case the keySet may not return values in
+        // order. header map is <key, position>, order by position first, then construct the key set
+        Set<String> headers = records.getHeaderMap()
+                .entrySet()
                 .stream()
-                .filter(entry -> !columnsToRedact.contains(entry)).toArray(String[]::new);
+                .sorted(Comparator.comparingInt(Map.Entry::getValue))
+                .filter(entry -> !columnsToRedact.contains(entry.getKey()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+
+        // case-insensitive headers
+        Set<String> headersCI = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+        headersCI.addAll(headers);
+
+        // precondition, everything expected to be pseudonymized must exist in the rest of columns
+        columnsToPseudonymize
+            .forEach(columnToPseudonymize ->
+                Preconditions.checkArgument(headersCI.contains(columnToPseudonymize), "Column %s to be pseudonymized not in file", columnToPseudonymize));
+
 
         try(ByteArrayOutputStream baos = new ByteArrayOutputStream(1024);
             PrintWriter printWriter = new PrintWriter(baos);
-            CSVPrinter printer = new CSVPrinter(printWriter, CSVFormat.DEFAULT.withHeader(header))){
-
-            columnsToPseudonymize
-                    .forEach(columnToPseudonymize ->
-                            Preconditions.checkArgument(records.getHeaderMap().containsKey(columnToPseudonymize), "Column %s to be pseudonymized not in file", columnToPseudonymize));
+            CSVPrinter printer = new CSVPrinter(printWriter, CSVFormat.DEFAULT
+                .withHeader(headers.toArray(new String[0])))) {
```

Comment:
Just personal taste, I'd prefer this way:

```suggestion
                .withHeader(Iterables.toArray(headers, String.class))))) {
```